docs/cli/ilab-model-serve.md

```diff
@@ -0,0 +1,101 @@
+# Design for `ilab model serve`
+
+## Background
+
+With the [request from the community](https://github.com/instructlab/instructlab/issues/1106) for `ilab` to serve different backends such as [vllm](https://docs.vllm.ai/en/stable/) and the [cli redesign](https://github.com/instructlab/dev-docs/pull/48), this design doc's purpose is to flesh out the behavior of the `ilab model serve` command.
+
+Specifically the design of the command, its flags/options, and the how a new serving backend will be tested within the project.
+
+## Design
+
+Since the subject of the `ilab model serve` command is a model, regardless of the format of the model, every command takes in the `--model` flag or uses its default value in the config.
+
+`ilab model serve` has a subcommand for each backend it can serve models with. As of this draft, the two backends `ilab` would serve with are `llama-cpp` and `vllm`.
+
+This would lead to the commands:
+
+- `ilab model serve llama-cpp`
```

Comment:
As mentioned earlier, this feels unnatural: `ilab model serve llama-cpp`, I would imagine something like `ilab model serve --backend llama-cpp` instead. If additional backends or more complex configurations are added, the command structure might become less clear. Using `--backend` is immediately clear that `--backend` is an option, making the command more self-explanatory.
When designing the CLI, I like to remind myself of the following:

```bash
program action --flag <option>
```

Here llama-cpp is not an action, it's an option. That's how I see things, also understand that people might think otherwise :)

Reply:
IMO either option works well no strong opinion from me :)

Reply:
@mairin I don't think the second bullet can apply given that we are going the route of generic flags and not re-using backend-specific flags. This is described here: https://github.com/instructlab/dev-docs/pull/81/files#r1631408310

Other arguments from me on why I don't like the positional argument in this scenario:

* clarity: I know the value that I'm passing to `--backend` is a backend, when I pass `llama-cpp`, I might have no clue what it is
* stability: once we introduce `ilab model serve llama-cpp` we are blocked from adding any other things than a backend. Using `--backend` removes that problem. What if one day we want to add another verb/noun to `ilab model serve <whatever>`.
* consistency: it's a common CLI design practice to use flags to specify options

Reply:
One point that helps me decide these choices --

For llamacpp vs vllm, are the rest of the --flags identical? If not I'd prefer separate commands vs a --backend flag. I would prefer only doing something like `--backend` if the rest of the flags apply equally to both (and any future backend)

Reply:
@leseb Of the arguments you mentioned the one that stuck out to me the most is stability, in the sense that the next subcommand after `ilab model serve` has to be a backend. I understand wanting the flexibility for that subcommand in the future, but I'd argue we see the need for it now so we should just do it. 

Also I don't think it's the end of the world if the possibilities for the next subcommand aren't consistent. In fact if you start at the `ilab model` level, the only commonality is that the next level of sub-commands are related to models. The same principle applies to `ilab model serve`; the subcommands are related to model serving.

Reply:
> @leseb Of the arguments you mentioned the one that stuck out to me the most is stability, in the sense that the next subcommand after `ilab model serve` has to be a backend. I understand wanting the flexibility for that subcommand in the future, but I'd argue we see the need for it now so we should just do it.

Honestly, this isn't very future-proof. Since we're starting from scratch, we have the opportunity to design with future growth in mind and enhance user experience clarity.

> 
> Also I don't think it's the end of the world if the possibilities for the next subcommand aren't consistent. In fact if you start at the `ilab model` level, the only commonality is that the next level of sub-commands are related to models. The same principle applies to `ilab model serve`; the subcommands are related to model serving.

Yes, but today we think about `ilab model serve backend.` What if we later realized that the backend was just one of the possible sub-commands? It's not the end of the world, but it will degrade the user experience.

Reply:
While I was originally neutral, and perhaps leaning towards the subcommand. The recent proposals on the introduction of profiles shifted my perspective quite a bit on this one. 

I argue that with this capability users won't need to (or want to) care which backend is used to serve their model, all they really care about is that there model get served in a way that works and is optimal for the system they are using, and in a way that leads to good ilab results. Even if they do change it, often they will have only one workable choice for their system and use-case. As an example, running the full hifi SDG won't be possible with llama.cpp (at least today), and likewise, vllm is a bit too resource hungry for smaller scale systems.

Based on the above I think we should view this selection as an advanced option / parameter as opposed to a routine subcommand. There's also a compatibility aspect where users get guided into specifying a particular backend (since it's prominent when a subcommand) they might think they need to do this, and we could start to see more scripts and automation tied to it, which would make it harder for us to swap in a different technology we might introduce in the future (e.g say a server optimized for a particular model type, or specific GPU type). Granted this is still a risk with having the option at all, just likely less as an advanced option. 

Finally, I think there's an opportunity where we can make the technology more accessible by presenting higher level, more generic options. For those that are very specific to a particular backend we probably wouldn't want to continually sync ever low-level option, and those could be primarily pass through via the `[--vllm|--llama-cpp-args]` mechanism in this proposal.

Thoughts? 

I know we want to get this in asap, and ultimately I think we can make both approaches work. But I thought i would give this one more check in light of other recent design decisions.